cpp/src/arrow/filesystem/s3fs_test.cc

```diff
@@ -590,6 +590,17 @@ class TestS3FS : public S3TestMixin {
     AssertObjectContents(client_.get(), "bucket", "somefile", "new data");
   }
 
+  void TestOpenOutputStreamCloseAsyncDestructor() {
+    std::shared_ptr<io::OutputStream> stream;
+    ASSERT_OK_AND_ASSIGN(stream, fs_->OpenOutputStream("bucket/somefile"));
+    ASSERT_OK(stream->Write("new data"));
+    // Destructor implicitly closes stream and completes the multipart upload.
+    auto closeAsyncFut = stream->CloseAsync();
```

Comment:
IIUC, the intent of this test is to verify that even if we call `stream.reset()` like this we will still have non-segfaulting `closeAsyncFut`. Please add a comment making it clear that for that purpose it doesn't matter whether flush is triggered asynchronously after CloseAsync or sychronously after stream.reset() (which could be confusing since this test doesn't ensure ordering between the two)